app/controllers/auth_controller.rb

```diff
@@ -20,15 +19,14 @@ class App < Sinatra::Application
     @user = User.new
     @existing_user = User.find_by(email: params['email'])
     unless @existing_user
-      begin
-        flash[:notice] = 'wrong login credentials!'
-        redirect '/login'
-      end
+      flash[:notice] = 'wrong login credentials!'
+      redirect '/login'
     end
     result = @user.confirm_password(params['password'], @existing_user)
```

Comment:
I don't think this method is right. It is supposed to be a class method if you are passing both params. The user instance in this case does nothing. Try making it a class method in the model and calling `User.confirm_password(params['password'], @existing_user)` or if you still want to maintain it being an instance variable; try a method called `authenticate` that just takes the params


so, in this case you will reduce the code to something like
```ruby
@user = User.find_by(email: params['email'])

if @user && @user.authenticate(params['password'])
  session[:user_id] = @user.id
  session[:username] = @user.username
  session[:admin] = @user.admin?
  redirect '/topics'
end